### PR TITLE
fix(physics): rebuild world on episode reset and apply force at center of mass

### LINKED ISSUE #42: Box2D broadphase try/catch patches suppress memory corruption errors
## Summary
Box2D broadphase method monkey patches use empty try/catch blocks that suppress actual memory corruption errors instead of fixing root causes.

## Severity
Major

## Confidence
92

## Files / Locations
- src/core/physics-engine.ts lines 38-98

## Root Cause
The broadphase collision detection methods are wrapped in empty try/catch blocks that silently ignore all exceptions including native heap corruption, invalid pointer access, and out of bounds memory errors.

## Why This Matters
Silent non-deterministic physics drift accumulates over long rollouts with no visible failure indicators. Simulation state becomes corrupted and produces invalid physics results while appearing to run normally. This causes unreproducible training failures and invalid experiment results.

## Evidence
Lines 62, 75, and 88 contain empty catch blocks: `catch (e) {}` that discard all exceptions.

## Reproduction / Conditions
Occurs randomly when broadphase collision detection encounters corrupted Box2D internal state. Most frequently seen after 1000+ consecutive steps.

## Suggested Fix
1. Remove empty try/catch patches
2. Fix the underlying Box2D broadphase memory corruption bugs
3. Add proper error handling and recovery logic
4. Fail fast when corruption is detected

## Related Follow-Up
- Add long running stability tests for 100k+ consecutive steps
- Enable Box2D debug assertions
- Add crash dump logging for physics exceptions

### LINKED ISSUE #146: Player movement forces apply at player position instead of center of mass causing unintended torque
## Summary
Player movement input forces are applied at the player's world position instead of the center of mass, causing unintended angular acceleration (spinning) that breaks physics determinism and matches neither bonk.io's behavior nor the environment's documentation.

## Severity
Major

## Confidence
100

## Files / Locations
- src/core/physics-engine.ts:518

## Root Cause
In PhysicsEngine.applyInput() line 518, Box2D ApplyForce() is called with the player's world position (`pos`) as the application point instead of the body's local center of mass.

Box2D `ApplyForce(force, point)` expects `point` to be in **world coordinates**. When you pass the body's world position:
- For perfectly centered origin bodies this would be correct
- However, after any rotation, the body's origin position and its actual center of mass diverge
- This causes a torque to be applied every time movement input is pressed, making players spin uncontrollably

This is a classic Box2D API misuse bug.

## Why This Matters
1. **Physics correctness**: The environment does not match bonk.io's actual physics
2. **Determinism**: Small floating point differences in rotation accumulate into large gameplay differences
3. **RL training reliability**: Agents learn to exploit this unintended torque instead of valid gameplay mechanics
4. **Debuggability**: Spinning behavior is hard to attribute correctly to this root cause

## Evidence
Line 508-518 in src/core/physics-engine.ts:
```typescript
const pos = body.GetPosition();
const force = this._tempForce;
force.x = 0;
force.y = 0;

if (input.left) force.x -= MOVE_FORCE;
if (input.right) force.x += MOVE_FORCE;
if (input.up) force.y -= MOVE_FORCE;
if (input.down) force.y += MOVE_FORCE;

body.ApplyForce(force, pos);
```

The correct Box2D pattern for applying force through the center of mass with no torque is either:
1. Use `body.GetWorldCenter()` instead of `body.GetPosition()`
2. Use `ApplyForceToCenter(force)` which exists for exactly this purpose

## Reproduction / Conditions
Occurs 100% of the time when any movement input is pressed and the player body has any non-zero rotation. The effect becomes more pronounced the longer an episode runs.

## Suggested Fix
Replace line 518 with either:
```typescript
// Correct option 1: Use dedicated method
body.ApplyForceToCenter(force);

// Correct option 2: Explicitly use world center of mass
body.ApplyForce(force, body.GetWorldCenter());
```

## Related Follow-Up
- All existing benchmarks will change slightly with this fix
- Training checkpoints that learned the spinning exploit will need retraining
- Add a test case verifying zero angular acceleration from pure movement inputs

### PR DESCRIPTION
**Pull request by Varuu-0**
# Fresh-world episode reset & center-of-mass force point

Fixes the Box2D broadphase corruption family and the movement-torque divergence from native Bonk.io physics, grounded in deobfuscation evidence (`docs/DEOBFUSCATION.md` §31, §5, §35.2).

## What changed

- **Fresh world per episode** — `BonkEnvironment` resets now call `PhysicsEngine.reset()` (which discards the old `b2World` and rebuilds it) instead of `destroyAllBodies()`, eliminating the stale-broadphase trigger behind the corruption crashes. Fixes #42, #49, #77, #97, #112, #119.
- **Removed exception-swallowing broadphase monkey patches** — the `b2BroadPhase.Query/TestOverlap/MoveProxy/DestroyProxy` try/catch guards are deleted so corrupted physics fails loudly instead of silently desyncing (root-cause fix for the same family).
- **Removed the unsafe `destroyAllBodies()` API** — callers now use `reset()`; the 160 KB/reset world-reuse optimization was the source of the corruption and is retired.
- **Force applied at center of mass** — player movement uses `body.GetWorldCenter()` instead of `body.GetPosition()`, matching the decoded native `ApplyForce`/`GetWorldCenter` indices (163/164). Eliminates unintended movement-induced torque on off-center discs. Fixes #146.
- **Dead-disc detachment** — newly dead discs are removed from the live world after each step while their final transforms remain observable, mirroring the native next-state rebuild and fixing a latent fixed-AABB broadphase corruption found in the 10,000-tick throughput test.
- **Snapshotted dead-disc final state (PR review follow-up)** — when a dead disc is detached its final transforms are snapshotted into a plain-object map and the destroyed body is removed from `playerBodies`, so observations never read the destroyed `b2Body`. OOB detection and detachment share a single per-tick pass over a snapshot copy of `playerBodies`, so map mutation during detach can never skip a not-yet-visited id.

## Commits

| Commit | Area |
|--------|------|
| `68de683` fix(physics): rebuild world on episode reset and apply force at center of mass | Core physics + regressions |
| `e7feb6c` fix(physics): snapshot dead-disc final state and single-pass death handling | PR #176 review follow-up |
| `0fdbd40` fix(physics): iterate death pass over player map snapshot and widen lethal-fall test budget | PR #176 re-review follow-up |

## Tests

Full suite passes (44 files, 1,115 tests passed, 19 skipped). New regression coverage:

- `tests/unit/physics-engine.test.ts` — movement through an offset center of mass produces zero angular velocity (failed before: `angularVel ≈ 0.603`).
- `tests/unit/physics-engine.test.ts` — dead-disc final transforms and velocity stay readable and stable across 5 post-death ticks, and the destroyed body leaves `playerBodies` on detach (fails under retained-entry behavior).
- `tests/integration/environment-edge-cases.test.ts` — distinct world per episode with listener/warm-start/PPM/bounds/config restoration, body/capzone/joint/grapple recreation, tick reset, and lethal contacts preserved across 75 repeated 71-body resets.
- `tests/integration/environment-edge-cases.test.ts` — dead-player observation stays frozen across 3 post-death steps through the real observation path (200-step death budget; suite re-run 5 times to confirm stability).
- `tests/perf/physics-throughput.test.ts` — 10,000-tick run no longer crashes the broadphase.
- Remaining `destroyAllBodies()` tests converted to `reset()`.

## Notes

- `npm run typecheck` fails repo-wide on `main` due to pre-existing config (tsconfig target `es2016` without the libs/`@types/node` that benchmarks and tests require); no errors are introduced by this branch — changed modules typecheck cleanly under an isolated ES2020/Node check.
- Docs updated: `docs/DEOBFUSCATION_FIX_TRACKER.md` (evidence + closure mapping), `docs/PERFORMANCE.md`, `docs/roadmap.md`, `benchmarks/runner.ts`.

### CODE REVIEW

File: src/core/physics-engine.ts [hunk 1]
```diff
@@ -885,6 +820,17 @@ export class PhysicsEngine {
...
     }
 
+    // The native client rebuilds only the next state's live discs. This port
+    // keeps final transforms queryable, but removes dead proxies so they cannot
+    // leave the fixed broadphase AABB during direct post-death engine ticks.
+    for (const [id, body] of this.playerBodies) {
+      if (this.playerAlive.get(id) || this.detachedPlayerBodies.has(id)) continue;
+      this.releaseGrapple(id);
+      this.pendingSwingDestroy.delete(id);
+      this.world.DestroyBody(body);
```
[the remark this answers is not in the document]
Comment (Varuu-0) [reply src/core/physics-engine.ts]: Fixed in e7feb6c. Dead discs now snapshot their final transforms into a plain-object map (detachedPlayerStates) inside detachPlayer() and are removed from playerBodies; getPlayerState() serves the snapshot for detached players, so observations no longer read the destroyed b2Body. The regression `releases the destroyed body from playerBodies on detach so observations never read it` pins the release and fails under the previous retained-entry behavior. Full suite: 1,115 passed, 19 skipped.

File: src/core/physics-engine.ts [hunk 2]
```diff
@@ -885,6 +820,17 @@ export class PhysicsEngine {
       }
     }
 
+    // The native client rebuilds only the next state's live discs. This port
+    // keeps final transforms queryable, but removes dead proxies so they cannot
+    // leave the fixed broadphase AABB during direct post-death engine ticks.
+    for (const [id, body] of this.playerBodies) {
```
[the remark this answers is not in the document]
Comment (Varuu-0) [reply src/core/physics-engine.ts]: Fixed in e7feb6c. The OOB/death detection loop and the detach scan are merged into a single pass over playerBodies; dead bodies are removed from the map at detach time, so the common no-death tick performs one walk instead of two full passes.

File: src/core/physics-engine.ts [hunk 3]
```diff
@@ -885,6 +820,17 @@ export class PhysicsEngine {
...
+      if (this.playerAlive.get(id) || this.detachedPlayerBodies.has(id)) continue;
+      this.releaseGrapple(id);
+      this.pendingSwingDestroy.delete(id);
+      this.world.DestroyBody(body);
+      this.detachedPlayerBodies.add(id);
```
[the remark this answers is not in the document]
Comment (Varuu-0) [reply src/core/physics-engine.ts]: Fixed in e7feb6c. Added two regressions pinning the frozen final-transform contract: `snapshots final transforms and velocity so they stay readable and stable across post-death ticks` (engine-level, 5 post-death ticks) and `freezes the dead player observation across post-death steps` (environment-level, 3 post-death steps through the real observation path).

File: src/core/physics-engine.ts [hunk 4]
```diff
@@ -872,16 +843,23 @@ export class PhysicsEngine {
...
+        }
+      }
 
-      const pos = body.GetPosition();
-      if (pos.x * pos.x + pos.y * pos.y > this.oobRadiusSquared) {
-        this.playerAlive.set(id, false);
-        this.playerDeathType.set(id, 4);
-        globalProfiler.increment('death_out_of_bounds');
+      if (!this.playerAlive.get(id)) {
+        this.detachPlayer(id, body);
```
[the remark this answers is not in the document]
Comment (Varuu-0) [reply src/core/physics-engine.ts:863]: Fixed in 0fdbd40. The merged OOB/death pass now iterates `Array.from(this.playerBodies)` (a snapshot copy), so deleting an entry during detachPlayer can never cause a Map iterator to silently skip a not-yet-visited id under future multi-key cleanup; the body would otherwise stay live in the world. Full suite: 1,115 passed, 19 skipped.

File: tests/integration/environment-edge-cases.test.ts [hunk 5]
```diff
@@ -727,6 +727,146 @@ describe('BonkEnvironment edge cases', () => {
...
+        },
+        bodies: [
+          { name: 'lethal', type: 'rect', x: 0, y: 200, width: 800, height: 30, static: true, isLethal: true },
+        ],
+      };
+      env = new BonkEnvironment({ mapData, numOpponents: 0, randomOpponent: false, maxTicks: 200 });
+      env.reset();
+
+      let deathObs: Observation | null = null;
+      for (let i = 0; i < 80; i++) {
```
[the remark this answers is not in the document]
Comment (Varuu-0) [reply tests/integration/environment-edge-cases.test.ts]: Fixed in 0fdbd40. The dead-player observation test now budgets 200 steps instead of 80, with a comment documenting the ~76-step death time (numOpponents: 0 vacuous allOpponentsDead short-circuits physics to every other step). The edge-case integration suite passed 5 consecutive runs; full suite: 1,115 passed, 19 skipped.
